File: UK_news_scraper/delivery_registry.py
from __future__ import annotations

import argparse
from datetime import datetime, timezone
import json
from pathlib import Path

from .runtime_lock import exclusive_lock


DEFAULT_REGISTRY = Path.home() / ".codex" / "automations" / "uk" / "sent_run_ids.json"


def claim_delivery(summary_path: str | Path, registry_path: str | Path = DEFAULT_REGISTRY) -> dict:
    summary = _read_json(Path(summary_path))
    delivery_id = summary["delivery_id"]
    registry_path = Path(registry_path)
    with exclusive_lock(registry_path.with_suffix(".lock"), wait_seconds=10):
        registry = _read_json(registry_path, default={})
        legacy_record = registry.get(summary["run_id"])
        if legacy_record:
            return {
                "claimed": False,
                "delivery_id": delivery_id,
                "existing": legacy_record,
                "reason": "legacy run_id already sent",
            }
        existing = registry.get(delivery_id)
        if existing and existing.get("state") in {"claimed", "sent"}:
            return {"claimed": False, "delivery_id": delivery_id, "existing": existing}
        registry[delivery_id] = {
            "state": "claimed",
            "run_id": summary["run_id"],
            "status": summary["status"],
            "data_fingerprint": summary["data_fingerprint"],
            "claimed_at": datetime.now(timezone.utc).isoformat(),
            "excel_path": summary["output_file"],
        }
        _atomic_write_json(registry_path, registry)
    return {"claimed": True, "delivery_id": delivery_id}


def complete_delivery(
    delivery_id: str,
    message_id: str,
    registry_path: str | Path = DEFAULT_REGISTRY,
) -> dict:
    registry_path = Path(registry_path)
    with exclusive_lock(registry_path.with_suffix(".lock"), wait_seconds=10):
        registry = _read_json(registry_path, default={})
        record = registry.get(delivery_id)
        if not record or record.get("state") != "claimed":
            raise ValueError(f"delivery 尚未 claim：{delivery_id}")
        record.update(
            {
                "state": "sent",
                "message_id": message_id,
                "sent_at": datetime.now(timezone.utc).isoformat(),
            }
        )
        _atomic_write_json(registry_path, registry)
    return record


def release_delivery(
    delivery_id: str,
    registry_path: str | Path = DEFAULT_REGISTRY,
) -> bool:
    registry_path = Path(registry_path)
    with exclusive_lock(registry_path.with_suffix(".lock"), wait_seconds=10):
        registry = _read_json(registry_path, default={})
        record = registry.get(delivery_id)
        if not record or record.get("state") != "claimed":
            return False
        del registry[delivery_id]
        _atomic_write_json(registry_path, registry)
    return True


def delivery_status(
    delivery_id: str | None = None,
    state: str | None = None,
    registry_path: str | Path = DEFAULT_REGISTRY,
) -> dict:
    registry = _read_json(Path(registry_path), default={})
    if delivery_id:
        return {
            "delivery_id": delivery_id,
            "record": registry.get(delivery_id),
        }
    records = {
        key: value
        for key, value in registry.items()
        if not state or value.get("state") == state
    }
    return {"count": len(records), "records": records}


def recover_claim(
    delivery_id: str,
    confirmed: bool,
    registry_path: str | Path = DEFAULT_REGISTRY,
) -> dict:
    if not confirmed:
        raise ValueError("recover 需要明確確認：--confirm-release")
    released = release_delivery(delivery_id, registry_path)
    return {"delivery_id": delivery_id, "released": released}


def _read_json(path: Path, default=None):
    if not path.exists():
        return default
    return json.loads(path.read_text(encoding="utf-8"))


def _atomic_write_json(path: Path, payload) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary_path = path.with_suffix(path.suffix + ".tmp")
    temporary_path.write_text(
        json.dumps(payload, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )
    temporary_path.replace(path)


def main() -> None:
    parser = argparse.ArgumentParser(description="原子管理 UK 新聞郵件 delivery claim。")
    parser.add_argument("--registry", default=str(DEFAULT_REGISTRY))
    subparsers = parser.add_subparsers(dest="command", required=True)
    claim = subparsers.add_parser("claim")
    claim.add_argument("--summary", required=True)
    complete = subparsers.add_parser("complete")
    complete.add_argument("--delivery-id", required=True)
    complete.add_argument("--message-id", required=True)
    release = subparsers.add_parser("release")
    release.add_argument("--delivery-id", required=True)
    status = subparsers.add_parser("status")
    status.add_argument("--delivery-id")
    status.add_argument("--state", choices=("claimed", "sent"))
    recover = subparsers.add_parser("recover")
    recover.add_argument("--delivery-id", required=True)
    recover.add_argument("--confirm-release", action="store_true")
    args = parser.parse_args()

    if args.command == "claim":
        result = claim_delivery(args.summary, args.registry)
    elif args.command == "complete":
        result = complete_delivery(args.delivery_id, args.message_id, args.registry)
    elif args.command == "release":
        result = {"released": release_delivery(args.delivery_id, args.registry)}
    elif args.command == "status":
        result = delivery_status(args.delivery_id, args.state, args.registry)
    else:
        result = recover_claim(args.delivery_id, args.confirm_release, args.registry)
    print(json.dumps(result, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: native/crates/uk-news-app/src/delivery_registry.rs
use anyhow::{Context, Result};
use chrono::Utc;
use fs2::FileExt;
use serde_json::{json, Map, Value};
use std::{
    fs::{self, File, OpenOptions},
    io::Write,
    path::{Path, PathBuf},
};

fn default_registry() -> PathBuf {
    dirs_path().join(".codex/automations/uk/sent_run_ids.json")
}
fn dirs_path() -> PathBuf {
    std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("."))
}
struct Lock(File);
impl Drop for Lock {
    fn drop(&mut self) {
        let _ = self.0.unlock();
    }
}
fn lock(path: &Path) -> Result<Lock> {
    if let Some(p) = path.parent() {
        fs::create_dir_all(p)?
    }
    let file = OpenOptions::new()
        .create(true)
        .truncate(false)
        .read(true)
        .write(true)
        .open(path)?;
    file.lock_exclusive()?;
    Ok(Lock(file))
}
fn read(path: &Path) -> Result<Map<String, Value>> {
    if !path.exists() {
        return Ok(Map::new());
    }
    serde_json::from_str::<Value>(&fs::read_to_string(path)?)?
        .as_object()
        .cloned()
        .context("delivery registry must be an object")
}
fn write(path: &Path, map: &Map<String, Value>) -> Result<()> {
    if let Some(p) = path.parent() {
        fs::create_dir_all(p)?
    }
    let tmp = path.with_extension("json.tmp");
    let mut f = File::create(&tmp)?;
    f.write_all(serde_json::to_string_pretty(map)?.as_bytes())?;
    f.write_all(b"\n")?;
    f.sync_all()?;
    fs::rename(tmp, path)?;
    Ok(())
}
pub fn claim(summary: &Path, registry: Option<&Path>) -> Result<Value> {
    let s: Value = serde_json::from_str(&fs::read_to_string(summary)?)?;
    for k in [
        "delivery_id",
        "run_id",
        "status",
        "data_fingerprint",
        "output_file",
    ] {
        if s.get(k)
            .and_then(Value::as_str)
            .filter(|x| !x.is_empty())
            .is_none()
        {
            anyhow::bail!("run summary 缺少欄位：{k}")
        }
    }
    let path = registry.map(Into::into).unwrap_or_else(default_registry);
    let _lock = lock(&path.with_extension("lock"))?;
    let mut map = read(&path)?;
    let id = s["delivery_id"].as_str().unwrap();
    let run = s["run_id"].as_str().unwrap();
    if let Some(existing) = map.get(run) {
        return Ok(
            json!({"claimed":false,"delivery_id":id,"existing":existing,"reason":"legacy run_id already sent"}),
        );
    }
    if let Some(existing) = map.get(id) {
        if matches!(
            existing.get("state").and_then(Value::as_str),
            Some("claimed" | "sent")
        ) {
            return Ok(json!({"claimed":false,"delivery_id":id,"existing":existing}));
        }
    }
    map.insert(id.into(),json!({"state":"claimed","run_id":run,"status":s["status"],"data_fingerprint":s["data_fingerprint"],"claimed_at":Utc::now().to_rfc3339(),"excel_path":s["output_file"],"message_id":"","sent_at":""}));
    write(&path, &map)?;
    Ok(json!({"claimed":true,"delivery_id":id}))
}
pub fn complete(id: &str, message: &str, registry: Option<&Path>) -> Result<Value> {
    let path = registry.map(Into::into).unwrap_or_else(default_registry);
    let _lock = lock(&path.with_extension("lock"))?;
    let mut map = read(&path)?;
    let item = map
        .get_mut(id)
        .and_then(Value::as_object_mut)
        .context(format!("delivery 尚未 claim：{id}"))?;
    if item.get("state").and_then(Value::as_str) != Some("claimed") {
        anyhow::bail!("delivery 尚未 claim：{id}")
    }
    item.insert("state".into(), json!("sent"));
    item.insert("message_id".into(), json!(message));
    item.insert("sent_at".into(), json!(Utc::now().to_rfc3339()));
    let result = Value::Object(item.clone());
    write(&path, &map)?;
    Ok(result)
}
pub fn release(id: &str, registry: Option<&Path>) -> Result<bool> {
    let path = registry.map(Into::into).unwrap_or_else(default_registry);
    let _lock = lock(&path.with_extension("lock"))?;
    let mut map = read(&path)?;
    let release = map
        .get(id)
        .and_then(|x| x.get("state"))
        .and_then(Value::as_str)
        == Some("claimed");
    if release {
        map.remove(id);
        write(&path, &map)?
    }
    Ok(release)
}
pub fn status(id: Option<&str>, state: Option<&str>, registry: Option<&Path>) -> Result<Value> {
    let path = registry.map(Into::into).unwrap_or_else(default_registry);
    let map = read(&path)?;
    if let Some(id) = id {
        return Ok(json!({"delivery_id":id,"record":map.get(id)}));
    }
    let records = map
        .into_iter()
        .filter(|(_, v)| state.is_none() || v.get("state").and_then(Value::as_str) == state)
        .collect::<Map<_, _>>();
    Ok(json!({"count":records.len(),"records":records}))
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn python_registry_shape_round_trips_claim_complete() {
        let dir = tempdir().unwrap();
        let registry = dir.path().join("sent_run_ids.json");
        let summary = dir.path().join("report.run.json");
        fs::write(&summary, r#"{"delivery_id":"run:complete:abc","run_id":"run","status":"complete","data_fingerprint":"abcdef","output_file":"/tmp/report.xlsx"}"#).unwrap();
        assert_eq!(claim(&summary, Some(&registry)).unwrap()["claimed"], true);
        let record = complete("run:complete:abc", "gmail-123", Some(&registry)).unwrap();
        assert_eq!(record["state"], "sent");
        assert_eq!(record["message_id"], "gmail-123");
        assert!(!release("run:complete:abc", Some(&registry)).unwrap());
    }

    #[test]
    fn release_is_idempotent_for_claimed_record() {
        let dir = tempdir().unwrap();
        let registry = dir.path().join("sent_run_ids.json");
        let summary = dir.path().join("report.run.json");
        fs::write(&summary, r#"{"delivery_id":"run:degraded:def","run_id":"run","status":"degraded","data_fingerprint":"def","output_file":"/tmp/report.xlsx"}"#).unwrap();
        claim(&summary, Some(&registry)).unwrap();
        assert!(release("run:degraded:def", Some(&registry)).unwrap());
        assert!(!release("run:degraded:def", Some(&registry)).unwrap());
    }
}
